app/src/main/java/ardjomand/leonardo/arcmovies/upcomingmovies/UpcomingMoviesFragment.java

```diff
@@ -0,0 +1,179 @@
+import android.view.ViewGroup;
+import android.widget.Toast;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import ardjomand.leonardo.arcmovies.R;
+import ardjomand.leonardo.arcmovies.data.MoviesRepositoryImpl;
+import ardjomand.leonardo.arcmovies.model.UpcomingMovie;
+
+/**
+ * A fragment representing a list of Items.
+ * <p/>
+ * Activities containing this fragment MUST implement the {@link OnUpcomingMoviesFragmentListener}
+ * interface.
+ */
+public class UpcomingMoviesFragment extends Fragment implements UpcomingMoviesContract.View {
+
+    //region Constants
+    private static final String ARG_COLUMN_COUNT = "column-count";
+    private static final double PAGINATION_THRESHOLD = 5;
+    private static final String LIST_DATA = "List_Data";
+    private static final String LOG_LIST_STATE = "List_state";
+    //endregion
+
+    //region Member Variables
+    private int mColumnCount = 2;
+    private OnUpcomingMoviesFragmentListener mListener;
+    private UpcomingMoviesContract.Presenter mPresenter;
+    private UpcomingMoviesAdapter mAdapter;
+    private List<UpcomingMovie> mUpcomingMovies = new ArrayList<>();
+    //endregion
+
+    //region Constructors
+    public UpcomingMoviesFragment() {
+        // Required empty public constructor
+    }
+
+    public static UpcomingMoviesFragment newInstance(int columnCount) {
+        UpcomingMoviesFragment fragment = new UpcomingMoviesFragment();
+        Bundle args = new Bundle();
+        args.putInt(ARG_COLUMN_COUNT, columnCount);
+        fragment.setArguments(args);
+        return fragment;
+    }
+    //endregion
+
+    //region Lifecycle
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setRetainInstance(true);
+
+        if (getArguments() != null) {
+            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
+        }
+
+        new UpcomingMoviesPresenter(this, new MoviesRepositoryImpl());
+    }
+
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_upcoming_movies_list, container, false);
+
+        // Setup the RecyclerView
+        if (view instanceof RecyclerView) {
+            Context context = view.getContext();
+            RecyclerView recyclerView = (RecyclerView) view;
+
+            mAdapter = new UpcomingMoviesAdapter(getContext(), new ArrayList<UpcomingMovie>(), mListener);
+            recyclerView.setAdapter(mAdapter);
+
+            final GridLayoutManager gridLayoutManager = new GridLayoutManager(context, mColumnCount);
+            recyclerView.setLayoutManager(gridLayoutManager);
+
+            // Setup pagination
+            recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
+                @Override
+                public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
```

Comment:
This is called *many* times when scrolling. It is specially visible if the internet is out (many errors). How to avoid the extra calls when more movies are being loaded but haven't arrived yet?

Reply:
One possibility I considered while making the app was to have a isLoading member variable in the Fragment that would be checked before downloading more movies.
But that variable is already available on the Presenter and I thought duplicating it on the View would not be a very good idea.
So, even if the View makes a lot of calls to the Presenter, the Presenter only asks the Repository to download more data when its isLoading member variable is false.

Anyway, this does not fix the case when the network is not available. The simplest idea I can think of is to check for network connectivity inside the View before calling the Presenter.

I added a fix: when there is a failure getting data from the network, the mCurrentPage should be decremented again, otherwise when the connection returns some pages will be skipped.